src/ObservationDims.jl

```diff
@@ -1,5 +1,159 @@
+end
+
+## Vectors: obsdim is optional, we may or may not need it.
+for T in (Any, AbstractVector)
+
+    # Iterator -> IteratorOfObs
+    @eval organise_obs(::IteratorOfObs, obs_iter::$T; obsdim=nothing) = obs_iter
+
+    # Iterator -> ArraySlicesOfObs
+    @eval function organise_obs(::ArraySlicesOfObs{D}, obs_iter::$T; obsdim=nothing) where D
+
+        # we assume all obs have same number of dimensions else nothing makes sense
+        ndims_per_obs = ndims(first(obs_iter))
+
+        # If collection of scalars then return the collection
+        if ndims_per_obs == 0
+            return collect(obs_iter)
+        end
+
+        # This should just be a mapreduce but that is slow
+        # see https://github.com/JuliaLang/julia/issues/31137
+        shaped_obs = Base.Generator(obs_iter) do obs
+            new_shape = ntuple(ndims_per_obs + 1) do ii
+                if ii < D
+                    size(obs, ii)
+                elseif ii > D
+                    size(obs, ii-1)
+                else  # ii = D
+                    1
+                end
+            end
+            # add singleton dim that we wil concatenate on
+            Base.ReshapedArray(obs, new_shape, ())
+        end
+        return cat(shaped_obs...; dims=D)
+    end
+end
+
+# Specify arrangement based on desired ObsArrangement:
+
+# Any -> SingleObs: never any need to rearrage
+organise_obs(::SingleObs, data; obsdim=nothing) = data
+
+# If data is an array we can safely drop single length dimensions
+# This lets us compute metrics on pairs of objects that are (3,) and (3, 1)
+function organise_obs(::SingleObs, data::AbstractArray; obsdim=nothing)
+    to_drop = findall(size(data) .== 1)
+    data = isempty(to_drop) ? data : dropdims(data; dims=Tuple(to_drop))
+    return data
+end
+
+
+# Array -> IteratorOfObs or ArraySlicesOfObs: depends on obsdim
+# Resorts to default obsdim which redispatches to the 3 arg form below.
+for A in (IteratorOfObs, ArraySlicesOfObs)
+
+    @eval function organise_obs(arrangement::$A, data::AbstractArray; obsdim=_default_obsdim(data))
+        return organise_obs(arrangement, data, obsdim)
+    end
+
+    @eval function organise_obs(arrangement::$A, data::NamedDimsArray; obsdim=_default_obsdim(data))
+        obsdim = (obsdim isa Symbol) ? NamedDims.dim(data, obsdim) : obsdim
+        return organise_obs(arrangement, data, obsdim)
+    end
+
+    @eval function organise_obs(arrangement::$A, data::AxisArray; obsdim=_default_obsdim(data))
+        obsdim = (obsdim isa Symbol) ? axisdim(data, Axis{obsdim}) : obsdim
+        return organise_obs(arrangement, data, obsdim)
+    end
+end
+
+# 3 arg forms rearrange (non 1D) arrays according to the obsdim
+
+# Slice up the array to get an iterator of observations
+function organise_obs(::IteratorOfObs, data::AbstractArray, obsdim::Integer)
+    # This is basically eachslice from julia 1.1+
+    return (selectdim(data, obsdim, ii) for ii in Base.axes(data, obsdim))
+end
```

Comment:
We should open an issue to just switch to `eachslice`? it's in Compat.jl https://github.com/JuliaLang/Compat.jl/blob/601e535df78d1cdade4f6eed76348a8bf36cf864/src/Compat.jl#L15-L27

Reply:
Or just go ahead and switch to eachslice (and add it to the dependencies)

Reply:
I added `Compat`